lib/api/_loaders/_base-loader.js

```diff
@@ -237,47 +238,122 @@ class BaseLoader extends EventEmitter {
 
   validateMethod() {}
 
-  defineArgs(parent) {
+  defineArgs(parent, namespacedApi) {
     const {commandName} = this;
     const commandFn = this.commandFn.bind(this);
+
+    let stringifiedNamespace;
+    let mainNamespace;
+    let apiToReturn;
+
+    if (this.namespace && Utils.isString(this.namespace)) {
+      stringifiedNamespace = this.namespace;
+      mainNamespace = this.namespace;
+    } else if (Array.isArray(this.namespace) && this.namespace.length > 0) {
+      stringifiedNamespace = this.namespace.join('.');
+      mainNamespace = this.namespace[0];
+    }
+
+    if (namespacedApi && mainNamespace) {
+      namespacedApi[mainNamespace] = namespacedApi[mainNamespace] || {};
+      apiToReturn = new Proxy(namespacedApi[mainNamespace], {
+        get(_, name) {
+          // namespacedApi[mainNamespace] may change after API loading is finished.
+          return namespacedApi[mainNamespace][name];
+        }
+      });
+    }
+
     const args = [this.createQueuedCommandFn({
       parent,
       commandName,
+      namespace: stringifiedNamespace,
       commandFn,
+      apiToReturn,
       context: this
     })];
 
-    const namespace = this.getTargetNamespace(parent);
+    const namespace = this.getTargetNamespace(parent, namespacedApi);
     if (namespace) {
       args.unshift(namespace);
     }
 
     return args;
   }
 
+  isRestrictedNamespace(parent, namespace) {
+    if (!parent) {
+      return false;
+    }
+
+    // do not load these namespaces on parent (page objects and within-context).
+    const namespacesRestrictedForParent = ['client'];
+
+    const isRestrictedNamespace = namespace
+      && namespacesRestrictedForParent.some((ns) => namespace.toString().startsWith(ns));
+
+    return isRestrictedNamespace;
+  }
+
   define(parent = null) {
     if (!this.commandFn) {
       return this;
     }
 
-    this.validateMethod(parent);
+    if (this.isRestrictedNamespace(parent, this.namespace)) {
+      return this;
+    }
 
     const {commandName, nightwatchInstance} = this;
-    const args = this.defineArgs(parent);
 
+    this.validateMethod(parent);
+    const args = this.defineArgs(parent);
     nightwatchInstance.setApiMethod(commandName, ...args);
+
     if (this.module && this.module.AliasName) {
-      nightwatchInstance.setApiMethod(this.module.AliasName, ...args);
+      this.commandName = this.module.AliasName;
+
+      this.validateMethod(parent);
```

Comment:
why do we need to call `validateMethod` twice?

Reply:
Because we should also make sure that the `AliasName` does not conflict with any other existing command.

Reply:
okay